File: app/src/main/java/remix/myplayer/request/LibraryUriRequest.java
package remix.myplayer.request;

import android.graphics.drawable.Animatable;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.facebook.drawee.backends.pipeline.Fresco;
import com.facebook.drawee.controller.ControllerListener;
import com.facebook.drawee.interfaces.DraweeController;
import com.facebook.drawee.view.SimpleDraweeView;
import com.facebook.imagepipeline.common.ResizeOptions;
import com.facebook.imagepipeline.image.ImageInfo;
import com.facebook.imagepipeline.request.ImageRequestBuilder;

import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import remix.myplayer.request.network.RxUtil;
import remix.myplayer.util.LogUtil;

/**
 * Created by Remix on 2017/12/4.
 */

public class LibraryUriRequest extends ImageUriRequest<String> {
    private static final String TAG = LibraryUriRequest.class.getSimpleName();
    protected SimpleDraweeView mImage;

    UriRequest mRequest;

    public LibraryUriRequest(@NonNull SimpleDraweeView image, @NonNull UriRequest request, RequestConfig config) {
        super(config);
        mImage = image;
        mRequest = request;
    }

    public void onError(String errMsg) {
//        mImage.setImageURI(Uri.EMPTY);
        LogUtil.i(TAG, "onError: " + errMsg);
    }

    public void onSuccess(String result) {
        if (TextUtils.isEmpty(result)) {
            onError("empty result");
            return;
        }
        LogUtil.i(TAG, "onSuccess: " + result);
        ImageRequestBuilder imageRequestBuilder = ImageRequestBuilder.newBuilderWithSource(Uri.parse(result));
        if (mConfig.isResize()) {
            imageRequestBuilder.setResizeOptions(ResizeOptions.forDimensions(mConfig.getWidth(), mConfig.getHeight()));
        }
        DraweeController controller = Fresco.newDraweeControllerBuilder()
                .setImageRequest(imageRequestBuilder.build())
                .setOldController(mImage.getController())
                .setControllerListener(new ControllerListener<ImageInfo>() {
                    @Override
                    public void onSubmit(String s, Object o) {

                    }

                    @Override
                    public void onFinalImageSet(String s, @Nullable ImageInfo imageInfo, @Nullable Animatable animatable) {
                    }

                    @Override
                    public void onIntermediateImageSet(String s, @Nullable ImageInfo imageInfo) {
                    }

                    @Override
                    public void onIntermediateImageFailed(String s, Throwable throwable) {
                        LogUtil.d(TAG, "onIntermediateImageFailed: " + throwable);
                    }

                    @Override
                    public void onFailure(String s, Throwable throwable) {
                        LogUtil.d(TAG, "onFailure: " + throwable);
                    }

                    @Override
                    public void onRelease(String s) {

                    }
                })
                .build();

        mImage.setController(controller);
    }

    @Override
    public Disposable load() {
        return getCoverObservable(mRequest)
                .compose(RxUtil.applyScheduler())
                .subscribeWith(new DisposableObserver<String>() {
                    @Override
                    protected void onStart() {
                        mImage.setImageURI(Uri.EMPTY);
                    }

                    @Override
                    public void onNext(String s) {
                        onSuccess(s);
                    }

                    @Override
                    public void onError(Throwable e) {
                        LibraryUriRequest.this.onError(e.toString());
                    }

                    @Override
                    public void onComplete() {

                    }
                });

    }
}

File: app/src/main/java/remix/myplayer/request/LibraryUriRequest.kt
package remix.myplayer.request

import android.graphics.drawable.Animatable
import android.net.Uri
import android.text.TextUtils
import com.facebook.drawee.backends.pipeline.Fresco
import com.facebook.drawee.controller.ControllerListener
import com.facebook.drawee.view.SimpleDraweeView
import com.facebook.imagepipeline.common.ResizeOptions
import com.facebook.imagepipeline.image.ImageInfo
import com.facebook.imagepipeline.request.ImageRequestBuilder
import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableObserver
import remix.myplayer.request.network.RxUtil
import timber.log.Timber
import java.lang.ref.WeakReference

/**
 * Created by Remix on 2017/12/4.
 */

open class LibraryUriRequest(image: SimpleDraweeView, protected val request: UriRequest,
                             config: RequestConfig) : ImageUriRequest<String>(config) {

  protected val ref = WeakReference(image)

  override fun onError(throwable: Throwable?) {
    Timber.v("onError, request: $request msg: ${throwable?.message}")
//    mImageRef.get()?.setImageURI(Uri.EMPTY)
  }

  override fun onSuccess(result: String?) {
    val imageRequestBuilder = ImageRequestBuilder
        .newBuilderWithSource(Uri.parse(result))
    if (mConfig.isResize) {
      imageRequestBuilder.resizeOptions = ResizeOptions.forDimensions(mConfig.width, mConfig.height)
    }

    val image = ref.get() ?: return
    val controller = Fresco.newDraweeControllerBuilder()
        .setImageRequest(imageRequestBuilder.build())
        .setOldController(image.controller)
        .setControllerListener(object : ControllerListener<ImageInfo> {
          override fun onSubmit(s: String?, o: Any?) {

          }

          override fun onFinalImageSet(s: String?, imageInfo: ImageInfo?,
                                       animatable: Animatable?) {
          }

          override fun onIntermediateImageSet(s: String?, imageInfo: ImageInfo?) {}

          override fun onIntermediateImageFailed(s: String?, throwable: Throwable?) {
          }

          override fun onFailure(s: String?, throwable: Throwable?) {
          }

          override fun onRelease(s: String?) {

          }
        })
        .build()

    image.controller = controller
  }

  override fun load(): Disposable {
    return getCoverObservable(request)
        .compose(RxUtil.applyScheduler())
        .subscribeWith(object : DisposableObserver<String>() {
          override fun onStart() {
            ref.get()?.setImageURI(Uri.EMPTY)
          }

          override fun onNext(s: String) {
            onSuccess(s)
          }

          override fun onError(e: Throwable) {
            this@LibraryUriRequest.onError(e)
          }

          override fun onComplete() {

          }
        })

  }

  companion object {
    private val TAG = LibraryUriRequest::class.java.simpleName
    const val ERROR_NO_RESULT = "empty result"
    const val ERROR_BLACKLIST = "blacklist"
  }
}
